### PR TITLE
op-service/pprof: Support profiling to file (#6739)

### LINKED ISSUE #6739: op-service: pprof util package improvements
- [ ] profile from start of program
  We should add the option to start profiling from the start of the program (with one of the profile types) to capture program start-up performance, and not have to call the pprof http API after startup.
- [ ] Use `cli.GenericFlag` to validate flag contents upon CLI flag parsing time.

### PR DESCRIPTION
**Pull request by ArtyomArtamonov**
**Description**

Added two new flags: 
- `pprof.type`: Selects which profile type to use. Allowed values are: `cpu`, `heap`, `goroutine`, `threadcreate`, `block`, `mutex`, `allocs`;
- `pprof.path`: Specifies the path where the pprof file will be saved. If it is directory, default name will be {pprof.type}.prof

The flag `pprof.enabled` can still be applied, of course, so these changes are backward-compatible.

A new utility service has been added — `oppprof.Service`. It includes a constructor with all the necessary flags, as well as `Start()` and `Stop(ctx)` methods. These methods are executed during the initialization and shutdown stages, respectively.

For this new flags to work, I added two generic flags: EnumFlag and PathFlag in `op-service/flags` module.

The following services have been equipped with a new feature:
- op-proposer
- op-node
- op-heartbeat
- op-challenger
- op-batcher

**Tests**

No new tests were added. It's challenging for me to write tests for this feature, but I'm open to suggestions.

**Additional context**

I noticed that op-node flags registration is a bit different from other services. I removed old flags that were responsible for profiling: `pprof.enabled`, `pprof.addr`, `pprof.port`, and injected them in the `init()` function (like other flags there).

**Metadata**

- Fixes #6739

### CODE REVIEW

**Comment by tynes** [comment]
Definitely supportive of this change

File: op-service/oppprof/cli.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+		},
+		&cli.IntFlag{
+			Name:    PortFlagName,
+			Usage:   "pprof listening port",
+			Value:   defaultListenPort,
+			EnvVars: opservice.PrefixEnvVar(envPrefix, "PPROF_PORT"),
+		},
+		&cli.GenericFlag{
+			Name:    ProfilePathFlagName,
+			Usage:   "pprof file path. If ends with directory, sets default filename {profileType}.prof",
```
Comment (trianglesphere) [inline op-service/oppprof/cli.go]:
```suggestion
			Usage:   "pprof file path. If it is a directory, the path is {dir}/{profileType}.prof",
```

File: op-service/flags/custom.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
+package flags
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+type EnumValue struct {
```
Comment (trianglesphere) [inline op-service/flags/custom.go]: We already have op-enum, is this duplicating it or does this do more?

File: op-service/oppprof/cli.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+	ListenEnabled bool
+	ListenAddr    string
+	ListenPort    int
+
+	ProfileType     string
+	ProfileDir      string
+	ProfileFilename string
+}
+
+func (m CLIConfig) Check() error {
```
Comment (trianglesphere) [inline op-service/oppprof/cli.go:110]: Do we need to check if there is a profile type & we don't have a profile dir/name? I guess if we don't have a dir, we'll just write a file in the same spot that the binary was run from

File: op-service/oppprof/service.go [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+}
+
+func (s *Service) startServer() error {
+	log.Debug("starting pprof server", "addr", net.JoinHostPort(s.listenAddr, strconv.Itoa(s.listenPort)))
+	mux := http.NewServeMux()
+
+	// have to do below to support multiple servers, since the
+	// pprof import only uses DefaultServeMux
+	mux.Handle("/debug/pprof/", http.HandlerFunc(httpPprof.Index))
+	mux.Handle("/debug/pprof/cmdline", http.HandlerFunc(httpPprof.Cmdline))
```
Comment (trianglesphere) [inline op-service/oppprof/service.go]: I see this is copied over, but I wonder if we should disable this option. We pass potentially sensitive values via the command line (most pertinently the L1 RPC url which could have an embedded API key, the L2 engine JWT secret is hidden b/c that value is passed in via a file).

**Review by trianglesphere — COMMENTED** [review]
I left some comments about specific behavior, but I like this new functionality.

File: op-service/oppprof/cli.go (hunk 1 above)
Comment (ArtyomArtamonov) [reply op-service/oppprof/cli.go]: Fixed this

File: op-service/flags/custom.go (hunk 2 above)
Comment (ArtyomArtamonov) [reply op-service/flags/custom.go]: It abstracts checking for the valid value, but I'm okay to stick with existing way. Removed it.

File: op-service/oppprof/cli.go (hunk 3 above)
Comment (ArtyomArtamonov) [reply op-service/oppprof/cli.go:110]: Your guess is right, but I thought this parameter should not be mandatory. We check if the provided directory exists in PathFlag, but only if it was provided. If you insist that we need to make it mandatory, I can fix that.

File: op-service/oppprof/service.go (hunk 4 above)
Comment (ArtyomArtamonov) [reply op-service/oppprof/service.go]: Point seems valid. But if we do, we break backwards compatibilty. It seems like a small change, but I can imagine someone relying on this. Should we create another small issue and mark that it breaks compatibility?

Comment (EvanJRichard) [reply op-service/oppprof/service.go]: A tough dilemma. IMO it's better to break backwards compatibility vs. allowing the potential "footgun" of accidentally leaking sensitive values. (edit: Best of all would be to somehow keep backwards compatibility and prevent the accidental leak, but it's not immediately obvious to me how we'd easily do that.)

File: op-node/node/node.go [hunk 5]
```diff
@@ -372,16 +370,19 @@ func (n *OpNode) initHeartbeat(cfg *Config) {
...
+	n.pprofService = oppprof.New(
+		cfg.Pprof.ListenEnabled,
+		cfg.Pprof.ListenAddr,
+		cfg.Pprof.ListenPort,
+		cfg.Pprof.ProfileType,
+		cfg.Pprof.ProfileDir,
+		cfg.Pprof.ProfileFilename,
+	)
+
+	if err := n.pprofService.Start(); err != nil {
```
[the remark this answers is not in the document]
Comment (EvanJRichard) [reply op-node/node/node.go:382]: I weakly recommend this suggested change, it seems good to wrap any possible `pprofService.Start()` error in a little more context

File: op-service/oppprof/service.go (hunk 4 above)
Comment (ArtyomArtamonov) [reply op-service/oppprof/service.go]: We can try to filter sensitive data via middleware. But it's risky if some new sensitive flag is added and we forgot to filter it out here. Removed this handler completely by now.

File: op-node/node/node.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (ArtyomArtamonov) [reply op-node/node/node.go:382]: I wrapped this error, but in most services there is some context added one call up. I changed error text a bit for profiling, and wrapped other errors when initializing op-challenger [here](https://github.com/ethereum-optimism/optimism/pull/8709/files#diff-efa7242835622c7b930ca3afee54dc7f3f40c69e1b2d869a395259551d1b6fe2R86)

File: op-service/oppprof/service.go (hunk 4 above)
Comment (ArtyomArtamonov) [reply op-service/oppprof/service.go]: Can you help, which branch should I select as target now? There are multiple release/ branches and Active Branches table in docs does not contain any of them.

Comment (trianglesphere) [reply op-service/oppprof/service.go]: Use develop as the target branch

**Review by EvanJRichard — APPROVED** [review]
Thanks for addressing comments. LGTM!

**Pull request merged**